ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -184,24 +189,59 @@ private void internalTerminalBlockReached(Bytes32 executionBlockHash) {
         // we are in pre-merge and terminal block is reached
         // try to obtain a payloadId now
         final Bytes32 terminalBlockHash = executionPayloadTerminalBlockHash.get();
-        final ForkChoiceState terminalForkChoiceState =
-            new ForkChoiceState(terminalBlockHash, terminalBlockHash, Bytes32.ZERO);
-        internalForkChoiceUpdated(terminalForkChoiceState);
-        checkState(
-            lastSentForkChoiceState.isPresent()
-                && lastSentForkChoiceState.get().equals(terminalForkChoiceState),
-            "Required fork choice state was not sent");
-        return internalGetPayloadId(parentBeaconBlockRoot, false);
+        return requestPayloadId(terminalBlockHash, Bytes32.ZERO, parentBeaconBlockRoot);
       }
     }
 
-    // Merge is complete so we must have a real payload, but we don't have one that matches
-    // TODO: try to obtain a payloadId on the fly?
+    // Merge is complete, so we must have a real payload, but we don't have one that matches
+
+    if (allowPayloadIdOnTheFlyRetrieval) {
+      // try to obtain a payloadId now
+      Bytes32 finalizedExecutionBlockHash;
+      try {
+        finalizedExecutionBlockHash =
+            lastSentForkChoiceState
+                .map(ForkChoiceState::getFinalizedBlockHash)
+                .orElseGet(this::retrieveFinalizedExecutionBlockHash);
+      } catch (Exception e) {
+        throw new IllegalStateException(
+            String.format(
+                "Error while trying to retrieve current finalized Execution Block Hash for Beacon Block Root %s",
+                parentBeaconBlockRoot),
+            e);
+      }
```

Comment:
It feels a bit weird to have a catch-all exception handler here.  Could we just make `retrieveFinalizedExecutionBlockHash` return an exception with a useful error message and let that propagate up instead?